src/ModelManager.ts

```diff
@@ -125,96 +147,181 @@ export class ModelManager {
-            initialModel = config.model;
+        if (!rootModelURL) {
+            throw new Error('Provide root model url to initialize ModelManager.');
         }
 
-        this._listenersMap = {};
-        this._fetchPromises = {};
-        this._initPromise = null;
-
-        const pageModelRoot = PathUtils.getMetaPropertyValue(MetaProperty.PAGE_MODEL_ROOT_URL);
-        const metaPropertyModelUrl = PathUtils.internalize(pageModelRoot);
-        const currentPathname = PathUtils.getCurrentPathname();
-        const sanitizedCurrentPathname = currentPathname ? PathUtils.sanitize(currentPathname) : '';
-
-        // Fetch the app root model
-        // 1. consider the provided page path
-        // 2. consider the meta property value
-        // 3. fallback to the model path contained in the URL
-        const rootModelURL = path || metaPropertyModelUrl || sanitizedCurrentPathname;
-        // @ts-ignore
-        const rootModelPath = PathUtils.sanitize(rootModelURL);
-
         if (!rootModelPath) {
             throw new Error('No root modelpath resolved! This should never happen.');
         }
 
-        if (!rootModelURL) {
-            throw new Error('ModelManager.js Cannot initialize without an URL to fetch the root model');
-        }
+        return this._initPromise;
+    }
 
-        if (!this._modelClient) {
-            this._modelClient = new ModelClient();
+    /**
+     * Initializes the ModelManager asynchronously using the given path to resolve a data model.
+     * Ideal use case would be for remote apps which do not need the page model to be passed down from the root.
+     * For remote apps with no model path, an empty store is initialized and data is fetched on demand by components.
+     *
+     * Once the initial model is loaded and if the data model doesn't contain the path of the current pathname,
+     * the library attempts to fetch a fragment of model.
+     *
+     * Root model path is resolved in the following order of preference:
+     * - page path provided via config
+     * - meta property: `cq:pagemodel_root_url`
+     * - current path of the page for default SPA
+     * - if none, it defaults to empty string
+     *
+     * @fires cq-pagemodel-loaded if root model path is available
+     */
+    public initializeAsync(config?: ModelManagerConfiguration | string): void {
+        this.destroy();
+
+        const modelConfig = this._toModelConfig(config);
+        const initialModel = modelConfig && modelConfig.model;
+
+        this._initializeFields(modelConfig);
+        this._getPathsForModel(modelConfig);
+
+        const { rootModelPath } = this._modelPaths;
+
+        this._modelStore = new ModelStore(rootModelPath, initialModel);
+
+        if (rootModelPath) {
+            this._setInitializationPromise(rootModelPath);
         }
+    }
+
+    /**
+     * Initializes the class fields for ModelManager
+     */
+    private _initializeFields(config?: ModelManagerConfiguration) {
```

Comment:
Isn't `ModelManager.ts` too big already? Do you see an opportunity to refactor the file into modules and sub-components?

Reply:
The file has been refactored quite a bit already. Can take up further refactoring as part of later optimization.